### PR TITLE
Daemon::Log::RotationType option added

### PR DESCRIPTION
**Pull request by pboguslawski**
OTRS Daemon uses own log rotation method which may create lot of files
in log directory (i.e. if DB is down for a while). This may be hard
to maintain, especially if system uses standard rotation stuff like
logrotate in Debian.

This mod introduces new SysConfig option Daemon::Log::RotationType
which allowes one to set OTRS Daemon log rotation method:
- `OTRS` - logs rotated by OTRS (default),
- `External` - OTRS will not rotate logs and external rotation
  mechanism (i.e. logrotate) should be used instead.

Sample logrotate configuration for OTRS Daemon
(/etc/logrotate.d/otrs) that may be used in Debian with
Daemon::Log::RotationType=External:

```
/opt/otrs/var/log/Daemon/*.log {
    missingok
    notifempty
    su otrs www-data
    create 660
    sharedscripts
    postrotate
        if /bin/su -c '/opt/otrs/bin/otrs.Daemon.pl status' -s /bin/bash otrs > /dev/null ; then \
            /bin/su -c '/opt/otrs/bin/otrs.Daemon.pl stop' -s /bin/bash otrs > /dev/null ; \
            /bin/su -c '/opt/otrs/bin/otrs.Daemon.pl start' -s /bin/bash otrs > /dev/null ; \
        fi;
    endscript
}
```

Related: https://dev.ib.pl/ib/otrs/issues/39
Author-Change-Id: IB#1050869

### CODE REVIEW

File: bin/otrs.Daemon.pl [hunk 1]
```diff
@@ -546,24 +546,28 @@ sub _PIDUnlock {
...
-    # backup old log files
-    use File::Copy qw(move);
-    if ( -e "$FileStdOut.log" ) {
-        move( "$FileStdOut.log", "$FileStdOut-$SystemTime.log" );
-    }
-    if ( -e "$FileStdErr.log" ) {
-        move( "$FileStdErr.log", "$FileStdErr-$SystemTime.log" );
+    # get log rotation type and backup old logs if logs should be rotated by OTRS
+    my $RotationType = $ConfigObject->Get('Daemon::Log::RotationType') || 'otrs';
+    $RotationType = lc $RotationType;
```
Comment (carlosfrodriguez) [inline bin/otrs.Daemon.pl]: This is not really needed

File: bin/otrs.Daemon.pl [hunk 2]
```diff
@@ -546,24 +546,28 @@ sub _PIDUnlock {
...
+    if ( $RotationType eq 'otrs' ) {
```
Comment (carlosfrodriguez) [inline bin/otrs.Daemon.pl]: This can be changed to 
```perl
if (lc $RotationType eq 'otrs' ) {
```

File: bin/otrs.Daemon.pl [hunk 3]
```diff
@@ -575,29 +579,31 @@ sub _LogFilesSet {
         open STDERR, '>>', "$FileStdErr.log";
     }
 
-    # remove not needed log files
-    my $DaysToKeep = $ConfigObject->Get('Daemon::Log::DaysToKeep') || 1;
-    my $DaysToKeepTime = $SystemTime - $DaysToKeep * 24 * 60 * 60;
+    # remove not needed log files if OTRS rotation is enabled
+    if ( $RotationType eq 'otrs' ) {
```
Comment (carlosfrodriguez) [inline bin/otrs.Daemon.pl]: Same as above

File: bin/otrs.Daemon.pl [hunk 4]
```diff
@@ -607,6 +613,11 @@ sub _LogFilesSet {
 sub _LogFilesCleanup {
     my %Param = @_;
 
+    # skip cleanup if OTRS log rotation is not enabled
+    my $RotationType = $Kernel::OM->Get('Kernel::Config')->Get('Daemon::Log::RotationType') || 'otrs';
+    $RotationType = lc $RotationType;
```
Comment (carlosfrodriguez) [inline bin/otrs.Daemon.pl]: This is not needed

File: bin/otrs.Daemon.pl [hunk 5]
```diff
@@ -607,6 +613,11 @@ sub _LogFilesSet {
...
+    return 1 if $RotationType ne 'otrs';
```
Comment (carlosfrodriguez) [inline bin/otrs.Daemon.pl]: This can be changed to
```perl
return 1 if lc $RotationType ne 'otrs';
```

**Comment by carlosfrodriguez** [comment]
Hi @pboguslawski 

Thank you for your pull request, unfortunately is can not be merged cleanly. please rebase it

Additionally, please take a look on the comments above.

I think this is a nice addition to the Daemon, I will be looking forward for your update,

((enjoy))
cr

**Comment by pboguslawski** [comment]
Hi,

Just added commits with recent changes from master merged (we don't rebase in public repos - see https://www.atlassian.com/git/tutorials/merging-vs-rebasing/conceptual-overview#the-golden-rule-of-rebasing) and simplified code you pointed.

Regards,
Pawel

**Comment by carlosfrodriguez** [comment]
Hi @pboguslawski,

This pull request was merged in https://github.com/OTRS/otrs/commit/b54ef72dd44ac4de20cd6943251be63f50e96f09, with small modifications.

((enjoy))
cr

**Pull request closed without merging**